src/monotonic_path_preview/mod.rs

```diff
@@ -0,0 +1,911 @@
+///
+/// [`Error::IoError`]: struct.Error.html#variant.IoError
+fn split_head_and_tail<P: AsRef<path::Path>>(
+    path: P,
+) -> Result<
+    (
+        path::PathBuf,
+        collections::VecDeque<ffi::OsString>,
+    ),
+    Error,
+> {
+    let path = path.as_ref();
+
+    debug!("Splitting head and tail of {:?}", path);
+
+    // The path's head is the prefix and root components (if any).
+    fn is_head_part(c: &path::Component) -> bool {
+        match c {
+            path::Component::Prefix(_) => true,
+            path::Component::RootDir => true,
+            _ => false,
+        }
+    }
+
+    let mut head: path::PathBuf = path.components()
+        .take_while(is_head_part)
+        .collect();
+
+    // If this path is purely relative, it's relative to our current
+    // directory.
+    if head.as_os_str() == "" {
+        head.push(".");
+    }
+
+    debug!("Raw head: {:?}", head);
+
+    // The path's head can be safely converted into an absolute
+    // path with .canonicalize(), since it must exist.
+    head = head.canonicalize()
+        .map_err(|err| Error::IoError {
+            err: err,
+            at: path.into(),
+        })?;
+
+    // The tail is kind of a queue of components to check. Since we will
+    // be adding and removing things from diffferent sources, we convert
+    // everything to an `OsString` to make the lifetimes easier.
+    let tail: collections::VecDeque<ffi::OsString> = path.components()
+        .skip_while(is_head_part)
+        .map(|each| each.as_os_str().to_os_string())
+        .collect();
+
+    debug!("Head: {:?}, tail: {:?}", head, tail);
+    Ok((head, tail))
+}
+
+/// Returns the target of the symlink at `path`, if it exists and is one.
+///
+/// If `path` is a symlink and we can read the target, returns
+/// `Ok(Some(target))`.
+/// If `path` is not a symlink, returns `Ok(None)`.
+/// If `path` does not exist, it's still not a symlink and this function
+/// returns `Ok(None)`.
+/// Otherwise, returns the relevant error.
+fn read_link_if_exists<P: AsRef<path::Path>>(
+    path: P,
+) -> Result<Option<path::PathBuf>, Error> {
+    let path = path.as_ref();
+
+    // In theory, we could just call .read_link() directly and handle the error
+    // result, but:
+    //
+    // - on Windows, the "can't read this because it isn't a symlink"
+    //   error is returned as an io::ErrorKind::Other rather than
+    //   io::ErrorKind::InvalidInput.
+    // - on Windows, some not-smart filesystems (like VirtualBox's shared
+    //   folders) don't understand the API call, and just return "invalid
+    //   operation".
+    //
+    // Therefore, we'll check if the thing is a symlink before trying to read
+    // it.
+    path.symlink_metadata()
+        // If we successfully got metadata, check if it's a symlink.
+        .map(|metadata| metadata.file_type().is_symlink())
+        // If we failed to get metadata...
+        .or_else(|err| {
```

Comment:
This kind of error handling is always confusing to me. Can this block be broken up into a few statements so it is easier to follow?

Maybe something like:
```
let path = PathArc::from(path);
let meta = path.symlink_metadata()?; // error conversion handled for you
if meta.is_symlink() {
    // ... stuff
} else {
    // other stuff
}
```